### PR TITLE
io.h: add Lock,Unlock aliases for compatibility

### LINKED ISSUE #12792: Backward compatibility for `Lock/Unlock` in `runtime/caml/io.h`
Channel locking was rewritten in #12446. It removed `Lock` and `Unlock` macros in favor to `caml_channel_lock` and `caml_channel_unlock` functions. 

This makes use of those functions/macros conditioned by the OCaml version. Is it possible to have a `Lock` and `Unlock` functions to keep backward compatibility and avoid conditional on OCaml version?

### PR DESCRIPTION
**Pull request by gasche**
fixes #12792

(cc @rjbou, @dra27, @xavierleroy)

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -474,7 +474,8 @@ Working version
   (Gabriel Scherer and Thomas Refis, review by Florian Angeletti
    and Vincent Laviron)
 
-- #12446: remove the hooks machinery around channel locking in runtimee/io.c
+- #12446, #12792: remove the hooks machinery around channel locking
+  in runtimee/io.c
```
Comment (dra27) [inline Changes]: While you're here 🙂
```suggestion
  in runtime/io.c
```

File: runtime/caml/io.h [hunk 2]
```diff
@@ -103,6 +103,11 @@ CAMLextern intnat caml_really_getblock (struct channel *, char *, intnat);
 CAMLextern void caml_channel_lock(struct channel *);
 CAMLextern void caml_channel_unlock(struct channel *);
 
+/* Lock and Unlock are compatibility aliases for OCaml<5.2.
+   Remove whenever 5.2 is old enough. (See #12792.) */
```
Comment (dra27) [inline runtime/caml/io.h:107]: For me, it's so much "old enough" as "whenever noone still targets both current OCaml and < 5.2" (and the deprecations in 5.0.0 suggest that if there's no reason not to support a wide range, code will........!). Perhaps:
```suggestion
/* Prior to OCaml 5.2, caml_channel_lock and caml_channel_unlock were hooks and
   Lock and Unlock provided a neater way of calling them in io.c. The macros are
   kept for third party code which manipulates channels in C, but targets both
   current OCaml and versions prior to 5.2 (see #12792.) */
```

**Review by dra27 — APPROVED** [review]
This is good for me - I agree that we're not under obligation to do this, but the cost is very small. A possible alternate comment.

(@xavierleroy should have a chance to comment one way or the other, rather than just taking my approval, please)

**Comment by xavierleroy** [comment]
I still think this is the wrong interpretation of `CAML_INTERNALS`, but I have other fights to fight.  Still: what about adding `CAML_DEPRECATED` to the compatibility macros?

**Comment by dra27** [comment]
Hmm, the problem from the user-perspective is that with the deprecated alert, that means one's 5.2- and 5.2+ supporting-code necessarily emits deprecation warnings on all the _recent_ versions of OCaml (i.e. one's code emits warnings when developing on the latest version of OCaml just because it also supports older ones). That would imply that we'd want to restore something like `compatibility.h` again, and that certainly feels very wrong for something which is `CAML_INTERNALS`, I agree.

We'd be been suggesting probes based on the version (`#if OCAML_MAJOR >= 5`, etc.) before, but maybe this is better to have to put in user-code?

```c
#define CAML_INTERNALS
#include <caml/io.h>

/* ... */

/* Locking machinery was simplified in OCaml 5.2 */
#ifndef Lock
#define Lock(channel) caml_channel_lock(channel)
#define Unlock(channel) caml_channel_unlock(channel)
#endif
```
?

**Comment by xavierleroy** [comment]
I agree the deprecation warning should not be added now, only later, for instance once 4.14 is past LTS and everyone can move to recent 5.x versions.

The `#ifndef Lock` trick is a good one.  I'll let you and @gasche decide where to put the backward compatibility cursor.

**Comment by gasche** [comment]
@dra27 has a neat solution for the situation where our users hit this incompatibility, figure out what the problem is, and then find this issue and use the solution advised here. But honestly I think that this is a waste of time for them, when there is something easy we can do to ensure they never see the problem in the first place, which is to click "merge" on the present PR. So I just did that.

File: runtime/caml/io.h (hunk 2 above)
Comment (gasche) [reply runtime/caml/io.h:107]: I was not fully confident that we need to be so detailed in our comment, given that "what's a good time to remove a compatibility alias" is a fairly generic question with always the same answer: "whenever no one really needs it anymore". I generally like to be detailed in comments, but I was too lazy to decide whether to integrate your comment or not, so I just went without as the path of least effort.

**Comment by rjbou** [comment]
Thanks!!

**Pull request merged**